SILGen: simplify cleanups and avoid critical edges.

atrick opened the pull request:
Keep the CFG in a valid, critical-edge free state that the SIL passes can handle.
Replace the ad-hoc on-the-fly cleanup block optimization with a more general block merging.

This simplifies SILGen now and will in subsequent commits allow massive simplification of SIL APIs and SIL passes.

Note that there's no downside to avoiding critical edges since they will be split anyway any various points in the SIL pipeline. However, there are many passes and utilities that can be made much simpler and more compute time and memory efficient.

atrick committed:
SILGen: Fix a variable name conflict.

atrick committed:
SILGen: Always emit the cleanup block.

This avoids critical edges, which will be split in the optimizer
anyway.

atrick committed:
Add SILBuilder constructors that take a DebugScope.

Allow a new SILBuilder to be created for an insertion point, while
providing all of its necessary context.

Ultimately, the builder's constructor should take an insertion point,
DebugLocation, and context. Then we won't need to pass SILLocation to
all of its methods.

This makes much more sense and is much safer than saving the insertion
point via an RAII object or defining a separate
SILBuilderWithScope. Those broken abstractions should go away.

atrick committed:
Add a SILGenBuilder constructor that takes a new insertion point.

This is the best way to inherit the provided SILGenBuilder's debug
scope and context, but insert at a different SIL location.

atrick committed:
SILGenFunction::mergeCleanupBlocks.

Avoid emitting unnecessary basic block for cleanup chains. This is a
general approach that handles all cases while simplifying SILGen
emission and keeping the CFG in a valid state during SILGen.

atrick committed:
Fix test cases for SILGen after removing critical edges.

atrick committed:
Add FIXME comments with suggestions to improve test case maintenance.

atrick committed:
Simplify SILGen/Condition.

This code was attempting to generate a critical edge on one side of a
condition. Not only is that undesirable, but it breaks the abstraction
and results an a lot of complexity.

As usual, fixing critical edges simply means removing unnecessary complexity.

gottesmm commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -4436,7 +4438,6 @@ RValue RValueEmitter::visitIfExpr(IfExpr *E, SGFContext C) {
     
     // FIXME: We could avoid imploding and reexploding tuples here.
     Condition cond = SGF.emitCondition(E->getCondExpr(),
-                                       /*hasFalse*/ true,
```
I think this commit needs to go before the previous one, no? Beyond that looks great!

gottesmm reviewed (commented):
Looks like you are dicing/splicing. I think I found some nice feedback. Want to give it to you before you slice and dice more.

gottesmm commented on lib/SILGen/SILGenFunction.h:
```diff
@@ -693,14 +691,15 @@ class LLVM_LIBRARY_VISIBILITY SILGenFunction
 
-  Condition emitCondition(SILValue V, SILLocation Loc, bool hasFalseCode = true,
-                          bool invertValue = false,
+  Condition emitCondition(SILValue V, SILLocation Loc, bool invertValue = false,
```
If that is always a true condition, could we put an assert into Condition's destructor to catch this?

gottesmm reviewed (commented):
One last one.

atrick commented on lib/SILGen/Condition.h:
```diff
@@ -51,29 +50,34 @@ class LLVM_LIBRARY_VISIBILITY Condition {
             SILLocation L)
     : TrueBB(TrueBB), FalseBB(FalseBB), ContBB(ContBB), Loc(L)
```
I added an assert.

atrick commented on lib/SILGen/Condition.cpp:
```diff
@@ -36,110 +31,22 @@ static SILLocation getContinuationLoc(SILBasicBlock &BB, SILLocation Fallback) {
-  // nothing to clean up after.
-  if (!ContBB) {
-    return SGF.B.getInsertionBB();
+  assert(!TrueBB && "enterTrue is always called.");
```
Yeah, I added some comments.

atrick commented:
@gottesmm thanks for looking at this. I'd like to add feedback from review as separate commits so you don't need to re-review anything, but in this case, you hadn't finished the first round, and rebasing was part of the feedback. So the commits are all new, but nothing has changed functionally. Unfortunately, github isn't listing the commits in the right order anymore--trust that they are in the right order on the branch. Incidentally, I think it's better to clone-for-review for anything non-trivial.

gottesmm reviewed (commented):
I still think 66bc68151dcfcc9c45869f08fab4220054adf12e should be before the previous one. What you would do is do the same change but make it so that you can pass false in that place. Then all usages of Condition pass false and we can eliminate that unfortunate code... no? I think together the end state is fine though.

Also:

1. I wanted to give a +1 on a particular comment. It reads awesome! 
2. Found a typo.

Beyond that LGTM.

atrick committed:
SILGen KeyPath: Create a basic block to split the critical edge.

This is the one place in SILGen where we need to explicitly split the
critical edge. createBasicBlockAndBranch is a simply utility for that.

atrick commented:
Rerunning benchmarks to see if the few changes were spurious. Locally, I see only improvements. In particular, CI reported a 10% regression on IterateData, but I see a 10% improvement (although it seems very flaky). The only reason for regression would be incidental change to inlining because the cost model doesn't handle unconditional branches sensibly.

For code size, there is a 0.1% decrease in the benchmark suite and  0.1% increase in the stdlib.

atrick commented:
@swift-ci benchmark.

atrick commented:
I see no measurable impact on compile time.

atrick commented:
@swift-ci test.

atrick commented:
@swift-ci test.

atrick commented:
@swift-ci test.

atrick commented:
@swift-ci test.
